compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -212,29 +222,335 @@ public void visit(BLangPackage pkgNode) {
+                updateClassClosureMap(classDef);
+            }
+        }
 
+        // Update function parameters.
+        for (BLangFunction function : pkgNode.functions) {
+            updateFunctionParams(function);
+        }
+        for (BLangTypeDefinition typeDef : pkgNode.typeDefinitions) {
+            if (typeDef.typeNode.getKind() == NodeKind.RECORD_TYPE) {
+                updateRecordInitFunction(typeDef);
+            }
+        }
         result = pkgNode;
     }
 
+    private boolean isObjectConstructorRelated(TopLevelNode pkgLevelNode) {
+
+        if ((pkgLevelNode.getKind() == NodeKind.CLASS_DEFN &&
+                ((BLangClassDefinition) pkgLevelNode).isObjectContructorDecl)) {
+            return true;
+        }
+        if (pkgLevelNode.getKind() == NodeKind.FUNCTION &&
+                ((BLangFunction) pkgLevelNode).flagSet.contains(Flag.OBJECT_CTOR)) {
+            return false;
+        }
+        return false;
+    }
+
+    private void createClosureMapUpdateExpression(BLangClassDefinition classDef,
+                                                  BVarSymbol blockMap,
+                                                  BVarSymbol classMapSymbol) {
+
+        OCEDynamicEnvironmentData oceEnvData = classDef.oceEnvData;
+        BVarSymbol oceMap = oceEnvData.mapBlockMapSymbol;
+
+        BLangFunction function = classDef.generatedInitFunction;
+
+        // self
+        BVarSymbol selfSymbol = function.receiver.symbol;
+        BLangSimpleVarRef selfVarRef = ASTBuilderUtil.createVariableRef(function.pos, selfSymbol);
+        // $map$block$_2
+        BLangSimpleVarRef refToBlockClosureMap = ASTBuilderUtil.createVariableRef(function.pos, blockMap);
+
+        // self.$map$objectCtor
+        BLangIdentifier identifierNode = ASTBuilderUtil.createIdentifier(function.pos, blockMap.name.value);
+        BLangFieldBasedAccess fieldAccess = ASTBuilderUtil.createFieldAccessExpr(selfVarRef, identifierNode);
+        fieldAccess.symbol = oceMap;
+        fieldAccess.setBType(classMapSymbol.type);
+        fieldAccess.expectedType = classMapSymbol.type;
+        fieldAccess.isStoreOnCreation = true;
+        fieldAccess.isLValue = true;
+        fieldAccess.fieldKind = FieldKind.SINGLE;
+        fieldAccess.leafNode = true;
+
+        // self.$map$objectCtor = $map$block$_2
+        BLangAssignment assignmentStmt = (BLangAssignment) TreeBuilder.createAssignmentNode();
+        assignmentStmt.expr = refToBlockClosureMap;
+        assignmentStmt.pos = function.pos;
+        assignmentStmt.setVariable(fieldAccess);
+        fieldAccess.parent = assignmentStmt;
+
+        BLangFunction generatedInitFunction = classDef.generatedInitFunction;
+        BLangBlockFunctionBody generatedInitFnBody = (BLangBlockFunctionBody) generatedInitFunction.body;
+
+        // self[$map$objectCtor$_<num1>] = $map$block$_<num2>
+        generatedInitFnBody.stmts.add(0, assignmentStmt);
+        assignmentStmt.parent = generatedInitFnBody;
```

Comment:
Do we need to set parent here? https://github.com/ballerina-platform/ballerina-lang/issues/34395